Feature/2019 12 20 ai ml electron app (#574)
ordina-jworks/ordina-jworks.github.io: _posts/2019-12-20-Our-Devoxx-Demo-App.md

Finally we call the update method which will be called for each update.

``` javascript
async function update(video, canvas, context, detector) {
    context.drawImage(video, 0, 0, canvas.width, canvas.height);

    const detectedClasses = await detector.detect(canvas);
    Loader.anotateCanvas(canvas, detectedClasses);
    updateList(detectedClasses);

    if (enableLiveUpdate) {
        setTimeout(update, 1000 / framerate, video, canvas, context, detector);
    }
}
```
This function will function as our update loop/tick.
We draw the image to the canvas, just as it was seen by the webcam.
We then pass the canvas into the detector which will detect any objects in the image data contained in the canvas.
This returns an array of detected classes which contain the type of objects, the score and a bounding box.
That information is fed to a utility method that will annotate the canvas with a red rectangle with the coordinates of the bounding and the type of object with its score.
The update function will be called again with a very basic (and very limited) framerate timeout.

> There is other code, mainly in the coco folder which is mostly taken from the [default google implementation](https://github.com/tensorflow/tfjs-models/tree/master/coco-ssd){:target="_blank" rel="noopener noreferrer"} and modified to match the needs of this application.

Reviewer: ```suggestion
There is other code, mainly in the `coco` folder which is mostly taken from the [default Google implementation](https://github.com/tensorflow/tfjs-models/tree/master/coco-ssd){:target="_blank" rel="noopener noreferrer"} and modified to match the needs of this application.
```

One thing worth noting is the changes required to a small piece of code to make the model work faster by allowing it to work asynchronously:
```typescript
    const [maxScores, classes] = this.calculateMaxScores(scores, result[0].shape[1], result[0].shape[2]);

    const indexTensor = tf.tidy(() => {
        const boxes2 = tf.tensor2d(boxes, [result[1].shape[1], result[1].shape[3]]);
        return tf.image.nonMaxSuppression(boxes2, maxScores, maxNumBoxes, 0.5, 0.5);
    });

    const indexes = indexTensor.dataSync() as Float32Array;
    indexTensor.dispose();
```

Needs to be changed into:

```typescript
    const [maxScores, classes] = this.calculateMaxScores(scores, result[0].shape[1], result[0].shape[2]);

    const boxes2 = tf.tensor2d(boxes, [result[1].shape[1], result[1].shape[3]]);
    const temp = await tf.image.nonMaxSuppressionAsync(boxes2, maxScores, maxNumBoxes, 0.5, 0.5);
    const indexTensor = tf.tidy(() => {
        return temp;
    });

    const indexes = indexTensor.dataSync() as Float32Array;
    indexTensor.dispose();
```

> Because the `tf.image.nonMaxSuppressionAsync`returns a Promise the replacement is not as simple as just replacing the method call to the async variant.

Reviewer: ```suggestion
Because the `tf.image.nonMaxSuppressionAsync`returns a `Promise`, the replacement is not as simple as just replacing the method call to the async variant.
```

> This however is the only big change that needs to be done (apart from changing some dependencies) to get the application working with full web-gl acceleration.

Reviewer: ```suggestion
This however is the only big change that needs to be done (apart from changing some dependencies) to get the application working with full WebGL acceleration.
```

> The application is actually really simple and very easy to understand and tinker with. I strongly encourage you to check out the repo and have a go getting it up and running by yourself.

Reviewer: ```suggestion
The application is actually really simple and very easy to understand and tinker with. I strongly encourage you to check out the repo and have a go at getting it up and running by yourself.
```

The other branches contain different solutions with some tweaked code, be sure to also check those out!

## Conclusion

Making a cool demo that utilizes machine learning and pre-trained models is not at all that hard. 
Rapid prototyping with these pre-trained models allows one to quickly see if a certain strategy or desired functionality is workable and merits further development effort.
It also provides a way to get started easily in a matter that is extremely hard to master. 

> Retraining or tweaking these models can be very hard and time consuming as it requires an in depth knowledge of the matter at hand (both mathematics and the actual data).

Reviewer: ```suggestion
Retraining or tweaking these models can be very hard and time consuming as it requires an in-depth knowledge of the matter at hand (both mathematics and the actual data).
```

Im my opinion these pre-trained models and other machine learning 'building blocks' provide an extremely valuable toolset for developers.

## Resources

> - [Demo github repo](https://github.com/ordina-jworks/devoxx-webcam-ml){:target="_blank" rel="noopener noreferrer"}

Reviewer: ```suggestion
- [Demo GitHub repo](https://github.com/ordina-jworks/devoxx-webcam-ml){:target="_blank" rel="noopener noreferrer"}
```
